common/src/main/java/org/ergoplatform/appkit/Eip4Token.java

```diff
@@ -0,0 +1,229 @@
+        byte[] assetType = getR7ByteArrayOrNull();
+        if (r7 == null) {
+            return AssetType.NONE;
+        } else if (assetType == null) {
+            return AssetType.UNKNOWN;
+        } else {
+            for (AssetType type : AssetType.values()) {
+                byte[] r7EntryForType = type.getR7ByteArrayForType();
+
+                if (r7EntryForType != null && Arrays.equals(r7EntryForType, assetType))
+                    return type;
+            }
+
+            return AssetType.UNKNOWN;
+        }
+    }
+
+    @Nullable
+    private byte[] getR7ByteArrayOrNull() {
+        // r7 specifies the asset type and should be Coll<Byte>
+        if (r7 != null && r7.getValue() instanceof Coll) {
+            Coll<?> assetType = (Coll<?>) r7.getValue();
+
+            if (assetType.length() > 0 && assetType.apply(0) instanceof Byte) {
+                return JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) r7.getValue());
+            } else {
+                return null;
+            }
+
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * @return true if this is an NFT token asset type according to EIP-4
+     */
+    public boolean isNftAssetType() {
+        byte[] assetType = getR7ByteArrayOrNull();
+        return (assetType != null && assetType.length == 2 && assetType[0] == 1);
+    }
+
+    /**
+     * @return Sha256 content hash for NFT types, or null for non-NFT types
+     */
+    @Nullable
+    public byte[] getNftContentHash() {
+        if (isNftAssetType()) {
+            return JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) r8.getValue());
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * @return content link for NFT types if available, otherwise null
+     */
+    @Nullable
+    public String getNftContentLink() {
+        if (r9 != null && isNftAssetType()) {
+            // if r9 ErgoValue is Coll[Byte], we have the direct link here
+            if (r9.getValue() instanceof Coll) {
+                return new String(JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) r9.getValue()),
+                    StandardCharsets.UTF_8);
+            } else if (r9.getValue() instanceof Tuple2) {
+                return new String(JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) ((Tuple2) r9.getValue())._1),
+                    StandardCharsets.UTF_8);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * @return cover image link for NFT types if available, otherwise null
+     */
+    @Nullable
+    public String getNftCoverImageLink() {
+        if (r9 != null && isNftAssetType() && r9.getValue() instanceof Tuple2) {
+            return new String(JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) ((Tuple2) r9.getValue())._2),
+                StandardCharsets.UTF_8);
+        } else {
+            return null;
+        }
+    }
+
+    @Nonnull
+    public ErgoValue<Coll<scala.Byte>> getMintingBoxR4() {
```

Comment:
I suggest to rename getNameAsErgoValue, and say in JavaDoc that one use case is to create minting box.
The same for the other methods

Reply:
The idea was that this name is unaware of the actual content, so registers can be filled without knowing anything about Eip4. If the name was changed, only code aware of the fact that R4 is used to encode the token name can fill the registers.

Reply:
But in EIP4 the meaning of the registers is fixed, so why not use the fixed meaning in the name?

Reply:
Two reasons:
* If we do so actual code wanting to fetch the registers (that is for example code for minting the tokens) need to know EIP4 and translate that the name is R4. This is exactly the opposite of encapsulation. Keep in mind that there is also a method for retrieving the token name, which is intended for callers that want to get the name without needing to know how and where it is stored
* Although it is unlikely, the meaning of R4 might be changed or extended in the future. This happened to R9 before, which once was the content link and later became a Pair of content links under certain circumstances. If this happens to R4, all code for minting tokens needs to be changed (which shows that encapsulation is violated)

Reply:
I don't get it. Eip4Token class is domain specific abstraction - a domain entity which has a set of properties.
Yes, those properties are mapped to registers, because it is how the data is encoded and represented. But this encoding can be hidden from dApps, which use Eip4Token class. (Or it is not the goal? Why?). 
The apps can manipulate with the registers via properties, and which registers are used for that is encapsulated.

Reply:
Yes, it is the goal to hide internals. To hide the internals from all outside views, we have two methods here:

* `Eip4Token.getTokenName()` is the method to call when a dApp want to retrieve the name of the token. It does not need to know where and how it is stored
* `Eip4Token.getMintingBoxR4()` is the method to call when a dApp wants to mint this token and needs to set register 4 of the minting box, without needing to know what is stored inside the register.

If we rename this method, there is no method available for setting the minting box registers without knowing the internals of what is stored inside the register.